Process_Library/GOCART2G_MieMod.F90

```diff
@@ -0,0 +1,563 @@
+!     RH input from the table is scaled 0 - 0.99
+!     We resolve the map to 0 - 0.990 in steps of 0.001 (991 total steps)
+      do j = 1, NRH_BINS
+         do i = this%nrh, 1, -1
+            if ((j-1) .ge. int(this%rh(i)*1000)) then
+               ip1 = i + 1
+               this%rhi(j) = i
+               if (ip1 .gt. this%nrh) then
+                  this%rha(j) = 0.
+               else
+                  this%rha(j) =  ( (j-1)/1000. - this%rh(i)) &
+                                /  ( this%rh(ip1)- this%rh(i))
+               endif
+               exit
+            endif
+         enddo
+!       print *, j, this%rhi(j), this%rha(j), this%rh(this%rhi(j))
+      enddo
+
+      return
+
+  contains
+
+     subroutine polint(x,y,n,xWant,yWant,yErr)
+       integer :: n
+       !  recall, table hard-wired single precision
+       real*8 :: x(n),y(n)
+       real   :: xWant, yWant, yErr
+
+       !  given array x(n) of independent variables and array y(n) of dependent
+       !  variables, compute the linear interpolated result yWant at xWant and return
+       !  with a dummy error estimate yErr.  Hacked up from Numerical Recipes Chapter 3
+
+       integer :: i, j
+       real    :: dx, slope
+       character(len=255) :: msg
+
+       !  on out of bounds, set i to lower or upper limit
+       i = 0
+       if(xWant .lt. x(1)) then
+       !   write(msg,*) "in polint, wanted: ", xWant, ", got lower bound: ", x(1)
+       !   call warn(myname,msg)
+       !    if (mapl_am_i_root()) print *,'in polint(), wnted: ', xWant, ', got lower bound: ', x(1)
+         i = 1
+       endif
+       if(xWant .gt. x(n)) then
+       !    write(msg,*) "in polint, wanted: ", xWant, ", got upper bound: ", x(n)
+       !    call warn(myname,msg)
+       !    if (mapl_am_i_root()) print *,'in polint(), wnted: ', xWant, ', got upper bound: ', x(n)
+
+         i = n
+       endif
+
+       !  if i is still zero find i less than xWant
+       if (i .eq. 0) then
+          do j = 1, n
+             if(xWant .ge. x(j)) i = j
+          enddo
+       endif
+
+       !  slope
+       if (i .eq. n) then
+          slope = 0.
+       else
+         slope = (y(i+1)-y(i)) / (x(i+1)-x(i))
+       endif
+       dx = xWant - x(i)
+       yWant = y(i) + slope*dx
+
+       yErr = 0.
+
+     end subroutine polint
+
+   end function GOCART2G_MieCreate
+   
+
+#define RANK_ 0
+#define BYWAVE_ 1
+#include "Query.H"
+#undef BYWAVE_
+
+#define RANK_ 1
+#define BYWAVE_ 1
+#include "Query.H"
+#undef BYWAVE_
+
+#define RANK_ 2
+#define BYWAVE_ 1
+#include "Query.H"
+#undef BYWAVE_
+
+#define RANK_ 3
+#define BYWAVE_ 1
+#include "Query.H"
+#undef BYWAVE_
+
+#define RANK_ 0
+#define BYCHANNEL_ 1
+#include "Query.H"
+#undef BYCHANNEL_
+
+#define RANK_ 1
+#define BYCHANNEL_ 1
+#include "Query.H"
+#undef BYCHANNEL_
+
+#define RANK_ 2
+#define BYCHANNEL_ 1
+#include "Query.H"
+#undef BYCHANNEL_
+
+#define RANK_ 3
+#define BYCHANNEL_ 1
+#include "Query.H"
+#undef BYCHANNEL_
```

Comment:
Further simplification can be done by collecting the "RANK" iteration inside one BYCHANNEL definition.   The same can be done with the BYWAVELENGTH section above.
```suggestion
#define BYCHANNEL_ 1

#define RANK_ 0
#include "Query.H"
#undef RANK_

#define RANK_ 1
#include "Query.H"
#undef RANK_

#define RANK_ 2
#include "Query.H"
#undef RANK_

#define RANK_ 3
#include "Query.H"
#undef RANK_

#undef BYCHANNEL_
```